maitake/src/wait/queue.rs

```diff
@@ -0,0 +1,927 @@
+    /// a stored wakeup when enqueueing.
+    ///
+    /// When in this state, the node is **not** part of the linked list, and
+    /// can be dropped without removing it from the list.
+    Start,
+
+    /// The waiter is waiting.
+    ///
+    /// When in this state, the node **is** part of the linked list. If the
+    /// node is dropped in this state, it **must** be removed from the list
+    /// before dropping it. Failure to ensure this will result in dangling
+    /// pointers in the linked list!
+    Waiting,
+
+    /// The waiter has been woken.
+    ///
+    /// When in this state, the node is **not** part of the linked list, and
+    /// can be dropped without removing it from the list.
+    Woken,
+}
+
+/// The queue's current state.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[repr(u8)]
+enum State {
+    /// No waiters are queued, and there is no pending notification.
+    /// Waiting while the queue is in this state will enqueue the waiter;
+    /// notifying while in this state will store a pending notification in the
+    /// queue, transitioning to [`State::Woken`].
+    Empty = 0b00,
+
+    /// There are one or more waiters in the queue. Waiting while
+    /// the queue is in this state will not transition the state. Waking while
+    /// in this state will wake the first waiter in the queue; if this empties
+    /// the queue, then the queue will transition to [`State::Empty`].
+    Waiting = 0b01,
+
+    /// The queue has a stored notification. Waiting while the queue
+    /// is in this state will consume the pending notification *without*
+    /// enqueueing the waiter and transition the queue to [`State::Empty`].
+    /// Waking while in this state will leave the queue in this state.
+    Woken = 0b10,
+
+    /// The queue is closed. Waiting while in this state will return
+    /// [`Closed`] without transitioning the queue's state.
+    ///
+    /// *Note*: This *must* correspond to all state bits being set, as it's set
+    /// via a [`fetch_or`].
+    ///
+    /// [`Closed`]: crate::wait::Closed
+    /// [`fetch_or`]: core::sync::atomic::AtomicUsize::fetch_or
+    Closed = 0b11,
+}
+
+#[derive(Clone, Debug)]
+enum Wakeup {
+    Empty,
+    Waiting(Waker),
+    One,
+    All,
+    Closed,
+}
+
+// === impl WaitQueue ===
+
+impl WaitQueue {
+    /// Returns a new `WaitQueue`.
+    #[must_use]
+    #[cfg(not(loom))]
+    pub const fn new() -> Self {
+        Self {
+            state: CachePadded::new(AtomicUsize::new(State::Empty.into_usize())),
+            queue: Mutex::new(List::new()),
+        }
+    }
+
+    /// Returns a new `WaitQueue`.
+    #[must_use]
+    #[cfg(loom)]
+    pub fn new() -> Self {
+        Self {
+            state: CachePadded::new(AtomicUsize::new(State::Empty.into_usize())),
+            queue: Mutex::new(List::new()),
+        }
+    }
+
+    /// Wake the next task in the queue.
+    ///
+    /// If the queue is empty, a wakeup is stored in the `WaitQueue`, and the
```

Comment:
Two things I don't grok:

What happens if I call "wake" a bunch with an empty queue?

How come this stores a "wakeup" in the queue, rather than keeping it at the "empty" state?

Reply:
it's necessary to store a single wakeup in the queue because a task may be attempting to enqueue itself *while* `wake` is called, which would result in a "lost" wake. multiple calls to `wake` on an empty queue will not store additional wakeups; storing a count of wakeups in the queue seems like a job for a semaphore, which i'd also like to add to `maitake`.